### PR TITLE
[confmap] Add featuregate to use stable expansion rules

### PR DESCRIPTION
**Pull request by TylerHelmuth**
#### Description
This PR adds a feature gate that will handle transitioning users away from expandconverter, specifically expanding `$VAR` syntax.

The wholistic strategy is:
  1. create a new feature gate, `confmap.unifyEnvVarExpansion`, that will be the single feature gate to manage unifying collector configuraiton resolution.
  2. Update expandconverter to return an error if the feature gate is enabled and it is about to expand `$VAR` syntax.
  3. Update `otelcol.NewCommand` to set a `DefaultScheme="env"` when the feature gate is enabled and no `DefaultScheme` is set, this handles `${VAR}` syntax.
  4. Separately, deprecate `expandconverter`.
  5. Follow a normal feature gate cycle.
  6. Removed the `confmap.unifyEnvVarExpansion`  feature gates and `expandconverter` at the same time

Supersedes https://github.com/open-telemetry/opentelemetry-collector/pull/10259

#### Link to tracking issue
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/10161
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/8215
Related to https://github.com/open-telemetry/opentelemetry-collector/issues/7111

#### Testing
Unit tests

### CODE REVIEW

File: confmap/converter/expandconverter/expand.go [hunk 1]
```diff
@@ -80,21 +80,27 @@ func (c converter) expandStringValues(value any) (any, error) {
 func (c converter) expandEnv(s string) (string, error) {
 	var err error
 	res := os.Expand(s, func(str string) string {
+		// This allows escaping environment variable substitution via $$, e.g.
+		// - $FOO will be substituted with env var FOO
+		// - $$FOO will be replaced with $FOO
+		// - $$$FOO will be replaced with $ + substituted env var FOO
+		if str == "$" {
+			return "$"
+		}
```
Comment (TylerHelmuth) [inline confmap/converter/expandconverter/expand.go:89]: Can handle this via https://github.com/open-telemetry/opentelemetry-collector/pull/10392 if you want

File: confmap/resolver.go [hunk 2]
```diff
@@ -108,7 +115,8 @@ func NewResolver(set ResolverSettings) (*Resolver, error) {
 
 	converters := make([]Converter, len(set.ConverterFactories))
 	for i, factory := range set.ConverterFactories {
-		converters[i] = factory.Create(set.ConverterSettings)
+		f := factory.Create(set.ConverterSettings)
+		converters[i] = f
```
Comment (mx-psi) [inline confmap/resolver.go:119]: Is this change needed?

File: confmap/converter/expandconverter/expand.go (hunk 1 above)
Comment (mx-psi) [reply confmap/converter/expandconverter/expand.go:90]: Merged!

File: confmap/resolver.go (hunk 2 above)
Comment (TylerHelmuth) [reply confmap/resolver.go:119]: No, that was from a previous attempt at something, I will revert

File: otelcol/command.go [hunk 3]
```diff
@@ -52,6 +53,11 @@ func updateSettingsUsingFlags(set *CollectorSettings, flags *flag.FlagSet) error
 	if len(resolverSet.URIs) == 0 {
 		return errors.New("at least one config flag must be provided")
 	}
+
+	if confmap.UseUnifiedEnvVarExpansionRules.IsEnabled() && set.ConfigProviderSettings.ResolverSettings.DefaultScheme == "" {
+		set.ConfigProviderSettings.ResolverSettings.DefaultScheme = "env"
```
Comment (TylerHelmuth) [inline otelcol/command.go:58]: This solution has a downside: it expects that the `ProviderFactories` list includes a provider with the `env` scheme. 

This will be true for all our own usage of `NewCommand`, but may not be true for custom usages.

File: confmap/resolver.go [hunk 4]
```diff
@@ -12,8 +12,15 @@ import (
 
 	"go.uber.org/multierr"
 	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/featuregate"
 )
 
+var UseUnifiedEnvVarExpansionRules = featuregate.GlobalRegistry().MustRegister("confmap.unifyEnvVarExpansion",
+	featuregate.StageAlpha,
+	featuregate.WithRegisterFromVersion("v0.102.0"),
```
Comment (codeboten) [inline confmap/resolver.go:21]:
```suggestion
	featuregate.WithRegisterFromVersion("v0.102.1"),
```

File: otelcol/command.go [hunk 5]
```diff
@@ -52,6 +53,11 @@ func updateSettingsUsingFlags(set *CollectorSettings, flags *flag.FlagSet) error
 	if len(resolverSet.URIs) == 0 {
 		return errors.New("at least one config flag must be provided")
 	}
+
+	if confmap.UseUnifiedEnvVarExpansionRules.IsEnabled() && set.ConfigProviderSettings.ResolverSettings.DefaultScheme == "" {
```
Comment (codeboten) [inline otelcol/command.go:57]: Can this flag be tested?

File: otelcol/command.go (hunk 3 above)
Comment (TylerHelmuth) [reply otelcol/command.go:58]: Since this is only affects users who enable the feature gate, and the feature gate is only useful if you're using the env provider, I believe we're safe.

File: confmap/resolver.go (hunk 4 above)
Comment (mx-psi) [reply confmap/resolver.go:21]: I guess v0.103.0 actually, no?

Comment (TylerHelmuth) [reply confmap/resolver.go]: Updated to 0.103.0

File: otelcol/command.go (hunk 5 above)
Comment (TylerHelmuth) [reply otelcol/command.go:57]: Added

**Review by codeboten — APPROVED** [review]

File: confmap/resolver.go [hunk 6]
```diff
@@ -12,8 +12,15 @@ import (
 
 	"go.uber.org/multierr"
 	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/featuregate"
 )
 
+var UseUnifiedEnvVarExpansionRules = featuregate.GlobalRegistry().MustRegister("confmap.unifyEnvVarExpansion",
```
Comment (mx-psi) [inline confmap/resolver.go:19]: Can we make it so that this is not a public symbol? (e.g. move to `confmap/internal`)

**Review by mx-psi — APPROVED** [review]
One small change and it's good to go from my side

**Review by evan-bradley — APPROVED** [review]

File: confmap/resolver.go (hunk 6 above)
Comment (TylerHelmuth) [reply confmap/resolver.go:19]: @mx-psi we cannot, since `otelcol.NewCommand` depends on it.

**Review by codeboten — COMMENTED** [review]
Needs rebasing

File: confmap/resolver.go (hunk 6 above)
Comment (TylerHelmuth) [reply confmap/resolver.go:19]: Can do
